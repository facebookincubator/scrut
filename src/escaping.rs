/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use anyhow::Context;
use anyhow::Result;
use clap::ValueEnum;
use unicode_categories::UnicodeCategories;

use crate::newline::BytesNewline;

/// Provide ASCII and unicode compatible strings with all non-printable
/// characters escaped
#[derive(Debug, Clone, ValueEnum)]
pub enum Escaper {
    /// All non ASCII and all non-printable ASCII characters are escaped
    Ascii,
    /// All non-printable Unicode characters are escaped
    Unicode,
}

impl Default for Escaper {
    fn default() -> Self {
        Self::Unicode
    }
}

impl Escaper {
    /// Returns provided byte sequence as string where all non-printable characters
    /// are escaped
    pub fn escaped_printable(&self, raw: &[u8]) -> String {
        match self {
            Escaper::Ascii => escaped_printable_ascii(raw),
            Escaper::Unicode => escaped_printable_unicode(raw),
        }
    }

    /// Returns provided byte sequence as string where all non-printable characters
    /// are escaped with the ` (escaped)` expectation marker added, when appropriate
    pub fn escaped_expectation(&self, raw: &[u8]) -> String {
        match self {
            Escaper::Ascii => escaped_expectation_ascii(raw),
            Escaper::Unicode => escaped_expectation_unicode(raw),
        }
    }

    pub fn has_unprintable(&self, raw: &[u8]) -> bool {
        match self {
            Escaper::Ascii => has_unprintable_ascii(raw),
            Escaper::Unicode => has_unprintable_unicode(raw),
        }
    }
}

/// Convenience wrapper for [`String::from_utf8_lossy`]
#[macro_export]
macro_rules! lossy_string {
    ($arg:expr) => {{ String::from_utf8_lossy($arg).to_string() }};
}

/// All non-printable bytes are rendered as hexadecimal escape sequence, all
/// white-spaces as escaped character classes
fn escaped_printable_ascii(bytes: &[u8]) -> String {
    if has_unprintable_ascii(bytes) {
        bytes.iter().map(byte_to_ascii).collect()
    } else {
        String::from_utf8_lossy(bytes).into()
    }
}

/// Returns whether given byte sequence contains characters from the non-printable
/// ASCII range
fn has_unprintable_ascii(bytes: &[u8]) -> bool {
    bytes.iter().any(|byte| !matches!(*byte, b'\x20'..=b'\x7e'))
}

fn byte_to_ascii(byte: &u8) -> String {
    match byte {
        // printable, but whitespace
        b'\n' => "\\n".to_string(),
        b'\r' => "\\r".to_string(),
        b'\t' => "\\t".to_string(),

        // printable, control
        //b'\x00' => "\\0".to_string(), <-- cannot be supported, as it conflicts with octal notation like "\077"
        b'\x07' => "\\a".to_string(),
        b'\x08' => "\\b".to_string(),
        b'\x0c' => "\\f".to_string(),
        b'\x0b' => "\\v".to_string(),

        // backslashes must now be escaped themselves
        b'\\' => "\\\\".to_string(),

        // printable character
        b'\x20'..=b'\x7e' => (*byte as char).to_string(),

        // .. the rest is NOT printable
        _ => format!("\\x{:02x}", *byte),
    }
}

/// Renders given line either with escape sequences (if it contains non-printable
/// characters) and denoted as `(escaped)` - or as-is.
fn escaped_expectation_ascii(line: &[u8]) -> String {
    let escaped = escaped_printable_ascii(line.trim_newlines());
    let encoded = lossy_string!(line.trim_newlines());
    if encoded == escaped {
        encoded
    } else {
        format!("{escaped} (escaped)")
    }
}

/// All non-printable unicode are rendered as hexadecimal escape sequence, all
/// white-spaces as escaped character classes, everything else is printed
fn escaped_printable_unicode(bytes: &[u8]) -> String {
    let mut seq = [0; 4];
    if let Ok(s) = String::from_utf8(bytes.to_vec()) {
        return s
            .chars()
            .map(|c| {
                if c.is_other() {
                    let raw = c.encode_utf8(&mut seq).as_bytes();
                    escaped_printable_ascii(raw)
                } else {
                    c.to_string()
                }
            })
            .collect::<Vec<_>>()
            .join("");
    }

    // fallback in case this is not really a unicode string, but a binary sequence
    escaped_printable_ascii(bytes)
}

/// Renders given line either with escape sequences (if it contains non-printable
/// characters) and denoted as `(escaped)` - or as-is.
fn escaped_expectation_unicode(line: &[u8]) -> String {
    let escaped = escaped_printable_unicode(line.trim_newlines());
    let encoded = lossy_string!(line.trim_newlines());
    if encoded == escaped {
        encoded
    } else {
        format!("{escaped} (escaped)")
    }
}

/// Returns whether given byte sequence contains unicode characters that are not
/// not printable
fn has_unprintable_unicode(bytes: &[u8]) -> bool {
    String::from_utf8(bytes.to_vec())
        .map(|s| s.chars().any(|c| c.is_other()))
        .unwrap_or(true)
}

pub fn strip_colors(input: &str) -> Result<String> {
    let stripped = strip_ansi_escapes::strip(input)
        .context("strip ansi escape sequences from rendered output")?;
    String::from_utf8(stripped).context("decode stripped bytes back to utf8 string")
}

#[cfg(test)]
mod tests {

    use super::escaped_printable_ascii;
    use super::escaped_printable_unicode;

    #[test]
    fn test_bytes_as_printable_unicode() {
        let tests = vec![
            ("foo", "foo"),
            ("foo \x1b[1mbar\x1b[0m", "foo \\x1b[1mbar\\x1b[0m"),
            ("foo ðŸ˜‚", "foo ðŸ˜‚"),
            ("fðŸ¦€o", "fðŸ¦€o"),
            ("foo\tbar", "foo\\tbar"),
            ("foo \x1b[1m\tðŸ˜‚\x1d[0m", "foo \\x1b[1m\\tðŸ˜‚\\x1d[0m"),
            (
                "foo \x1b[2;5;0;31;47m\tðŸ˜‚\x1d[0m",
                "foo \\x1b[2;5;0;31;47m\\tðŸ˜‚\\x1d[0m",
            ),
        ];

        for (from, expect) in tests {
            let escaped = escaped_printable_unicode(from.as_bytes());
            assert_eq!(*expect, escaped, "from `{from}`");
        }
    }

    #[test]
    fn test_bytes_as_printable_ascii() {
        let tests = vec![
            ("\x00\x01\x02", "\\x00\\x01\\x02"),
            ("foo", "foo"),
            ("foo \x1b[1mbar\x1b[0m", "foo \\x1b[1mbar\\x1b[0m"),
            ("foo ðŸ˜‚", "foo \\xf0\\x9f\\x98\\x82"),
            ("foo\tbar", "foo\\tbar"),
            (
                "foo \x1b[1m\tðŸ˜‚\x1d[0m",
                "foo \\x1b[1m\\t\\xf0\\x9f\\x98\\x82\\x1d[0m",
            ),
            (
                "foo \x1b[2;5;0;31;47m\tðŸ˜‚\x1d[0m",
                "foo \\x1b[2;5;0;31;47m\\t\\xf0\\x9f\\x98\\x82\\x1d[0m",
            ),
        ];

        for (from, expect) in tests {
            let escaped = escaped_printable_ascii(from.as_bytes());
            assert_eq!(expect, &escaped, "from `{from}`");
        }
    }
}
