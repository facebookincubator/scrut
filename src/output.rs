/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::fmt::Debug;
use std::fmt::Display;
use std::time::Duration;

use serde::ser::SerializeMap;
use serde::Serialize;

use crate::config::DEFAULT_SKIP_DOCUMENT_CODE;
use crate::escaping::Escaper;
use crate::formatln;
use crate::lossy_string;
use crate::newline::SplitLinesByNewline;

/// Product of a single execution that captures output and status
#[derive(Clone, PartialEq, Eq)]
pub struct Output {
    /// The STDERR output of the execution
    pub stderr: OutputStream,

    /// The STDOUT output of the execution
    pub stdout: OutputStream,

    /// The exit code the execution ended in. A value of `None` implies the
    /// execution did not return (i.e. aborted due to timeout)
    pub exit_code: ExitStatus,
}

impl Output {
    pub fn to_error_string(&self, escaper: &Escaper) -> String {
        let mut err = String::new();
        err.push_str(&formatln!("## STDOUT"));
        err.push_str(&self.stdout.to_output_string(Some("#> "), escaper));
        err.push_str(&formatln!("## STDERR"));
        err.push_str(&self.stderr.to_output_string(Some("#> "), escaper));
        err
    }
}

impl Debug for Output {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let escaper = &Escaper::default();
        let stdout = escaper
            .escaped_printable(&self.stdout.0)
            .replace("\\n", "\\n\n");
        let stderr = escaper
            .escaped_printable(&self.stderr.0)
            .replace("\\n", "\\n\n");

        write!(
            f,
            "# STDOUT\n{}\n# STDERR\n{}\n# EXITCODE: {}\n",
            stdout,
            stderr,
            &self.exit_code.to_string(),
        )
    }
}

impl Default for Output {
    fn default() -> Self {
        Self {
            stdout: vec![].into(),
            stderr: vec![].into(),
            exit_code: ExitStatus::Unknown,
        }
    }
}

impl Serialize for Output {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut map = serializer.serialize_map(Some(3))?;
        map.serialize_entry("exit_code", &self.exit_code.to_string())?;
        map.serialize_entry("stdout", &lossy_string!((&self.stdout).into()))?;
        map.serialize_entry("stderr", &lossy_string!((&self.stderr).into()))?;
        map.end()
    }
}

impl<T: ToString, U: ToString> From<(T, U, Option<i32>)> for Output {
    fn from(set: (T, U, Option<i32>)) -> Self {
        Self {
            stdout: OutputStream(set.0.to_string().into()),
            stderr: OutputStream(set.1.to_string().into()),
            exit_code: match set.2 {
                None => ExitStatus::Unknown,
                Some(code) => ExitStatus::Code(code),
            },
        }
    }
}

impl<T: ToString, U: ToString> From<(T, U)> for Output {
    fn from(set: (T, U)) -> Self {
        Output::from((set.0, set.1, Some(0)))
    }
}

impl From<Duration> for Output {
    fn from(timeout: Duration) -> Self {
        Self {
            stdout: vec![].into(),
            stderr: vec![].into(),
            exit_code: ExitStatus::Timeout(timeout),
        }
    }
}

impl From<ExitStatus> for Output {
    fn from(status: ExitStatus) -> Self {
        Self {
            stdout: vec![].into(),
            stderr: vec![].into(),
            exit_code: status,
        }
    }
}

/// The status an execution can finish in
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ExitStatus {
    /// Execution resulted in exit code
    Code(i32),

    /// Execution never finished due to timeout
    Timeout(Duration),

    /// Execution was skipped
    Skipped,

    /// Implies the execution was intentionally detached and any output should
    /// be ignored.
    Detached,

    /// Execution failed for unknown reason
    Unknown,
}

impl ExitStatus {
    /// Exit code 0 denotes success
    pub const SUCCESS: Self = Self::Code(0);

    /// Returns exit code as integer with -1 for timeout and -255 for unknown
    pub fn as_code(&self) -> i32 {
        match self {
            Self::Code(code) => *code,
            Self::Skipped => DEFAULT_SKIP_DOCUMENT_CODE,
            Self::Timeout(_) => -1,
            Self::Detached => -100,
            Self::Unknown => -255,
        }
    }
}

impl Display for ExitStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Code(code) => write!(f, "{}", code),
            Self::Timeout(duration) => write!(f, "timeout[{:.2}ms]", duration.as_millis()),
            Self::Skipped => write!(f, "skipped"),
            Self::Detached => write!(f, "detached"),
            Self::Unknown => write!(f, "unknown"),
        }
    }
}

impl From<i32> for ExitStatus {
    fn from(value: i32) -> Self {
        ExitStatus::Code(value)
    }
}

impl From<ExitStatus> for i32 {
    fn from(value: ExitStatus) -> Self {
        value.as_code()
    }
}

#[derive(Clone, PartialEq, Eq)]
pub struct OutputStream(Vec<u8>);

impl OutputStream {
    pub fn to_output_string(&self, prefix: Option<&str>, escaper: &Escaper) -> String {
        let prefix = prefix.unwrap_or("");
        let mut out = String::new();
        let bytes: &[u8] = self.into();
        let lines = bytes.split_at_newline();
        let ends_in_newline = !bytes.is_empty() && bytes[bytes.len() - 1] == b'\n';
        for (idx, line) in lines.iter().enumerate() {
            let expectation = escaper.escaped_expectation(line);
            let suffix = if !ends_in_newline
                && !expectation.ends_with(" (escaped)")
                && idx + 1 == lines.len()
            {
                " (no-eol)"
            } else {
                ""
            };
            out.push_str(&formatln!("{}{}{}", prefix, &expectation, suffix))
        }
        out
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        self.0.clone()
    }
}

impl From<Vec<u8>> for OutputStream {
    fn from(stream: Vec<u8>) -> Self {
        Self(stream)
    }
}

impl From<&[u8]> for OutputStream {
    fn from(stream: &[u8]) -> Self {
        Self(stream.to_vec())
    }
}

impl From<&str> for OutputStream {
    fn from(stream: &str) -> Self {
        Self(stream.as_bytes().to_vec())
    }
}

impl From<&OutputStream> for Vec<u8> {
    fn from(stream: &OutputStream) -> Self {
        stream.0.clone()
    }
}

impl<'a> From<&'a OutputStream> for &'a [u8] {
    fn from(stream: &'a OutputStream) -> Self {
        &stream.0[..]
    }
}

#[cfg(test)]
mod tests {
    use super::OutputStream;
    use crate::escaping::Escaper;

    #[test]
    fn test_output_stream_appends_no_eol() {
        let tests = vec![
            ("a", "a (no-eol)\n"),
            ("a\n", "a\n"),
            ("a\nb", "a\nb (no-eol)\n"),
            ("a\nb\n", "a\nb\n"),
        ];
        for (from, expect) in tests {
            let stream = OutputStream(from.as_bytes().to_vec());
            let to = stream.to_output_string(None, &Escaper::default());
            assert_eq!(expect, &to, "from input '{from}'");
        }
    }

    #[test]
    fn test_prefixed_output_stream() {
        let tests = vec![
            ("a\n", "> a\n"),
            ("a\nb\n", "> a\n> b\n"),
            ("a\nb\nc\n", "> a\n> b\n> c\n"),
        ];
        for (from, expect) in tests {
            let stream = OutputStream(from.as_bytes().to_vec());
            let to = stream.to_output_string(Some("> "), &Escaper::default());
            assert_eq!(expect, &to, "from input '{from}'");
        }
    }
}
