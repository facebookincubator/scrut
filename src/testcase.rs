use serde::ser::SerializeMap;
use serde::Serialize;
use serde::Serializer;

use crate::diff::Diff;
use crate::diff::DiffTool;
use crate::expectation::Expectation;
use crate::output::ExitStatus;
use crate::output::Output;

pub type Result<T> = anyhow::Result<T, TestCaseError>;

/// An aggregate that unifies all ingredients for a test: a title
/// of the expected and intended state of the world; what a specific
/// command line should output and why
#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct TestCase {
    /// A human readable description that clarifies the intention
    pub title: String,

    /// The valid shell expression that is to be executed
    pub shell_expression: String,

    /// The expectations that describe the output of the execution
    pub expectations: Vec<Expectation>,

    /// The expected exit code of the execution
    #[serde(serialize_with = "serialize_always_as_value")]
    pub exit_code: Option<i32>,

    /// The line number of this test in the original file (starting at 1)
    pub line_number: usize,
}

impl TestCase {
    /// Validate that the outcome of an execution matches with the assumed
    /// outcome in regards to exit code and (STDOUT) output, or return an
    /// [`TestCaseError`]
    pub fn validate(&self, output: &Output) -> Result<()> {
        if let ExitStatus::Code(exit_code) = output.exit_code {
            let expected = self.exit_code.unwrap_or(0);
            if exit_code != expected {
                return Err(TestCaseError::InvalidExitCode {
                    actual: exit_code,
                    expected,
                });
            }
        }
        let diff_tool = DiffTool::new(self.expectations.clone());
        let diff = diff_tool
            .diff((&output.stdout).into())
            .map_err(TestCaseError::InternalError)?;
        if diff.has_differences() {
            Err(TestCaseError::MalformedOutput(diff))
        } else {
            Ok(())
        }
    }

    pub(crate) fn shell_expression_lines(&self) -> usize {
        self.shell_expression.matches('\n').count() + 1
    }

    pub(crate) fn expectations_lines(&self) -> usize {
        self.expectations.len()
    }
}

fn serialize_always_as_value<S>(x: &Option<i32>, s: S) -> std::result::Result<S::Ok, S::Error>
where
    S: Serializer,
{
    s.serialize_i32(x.unwrap_or(0))
}

/// An error that occurs when the actual output of an execution does not
/// match with the expectations.
///
/// There are four causes why an error can be raised:
/// 1) MalformedOutput: A line of output does not match the expected content or form
/// 2) UnexpectedOutput: There are more lines of output than there are
///    expectations to validate the output. Hence the additional output is
///    unexpected.
/// 3) InsufficientOutput: There are more expectation than there is output. That
///    means some of the expectations could never be applied and must be
///    considered failed (assuming they are non-optional)
/// 4) InternalError: An error occurred during processing, e.g. invalid UTF8
#[derive(Debug)]
pub enum TestCaseError {
    /// The validation of the expectation for the given line failed (invalid input)
    MalformedOutput(Diff),

    /// An execution ends in an unexpected exit code
    InvalidExitCode { actual: i32, expected: i32 },

    /// Delegated internal errors, e.g. relating to decoding
    InternalError(anyhow::Error),

    /// Whether this test was skipped intentionally
    Skipped,
}

impl PartialEq for TestCaseError {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::MalformedOutput(l0), Self::MalformedOutput(r0)) => l0 == r0,
            (
                Self::InvalidExitCode {
                    actual: l_actual,
                    expected: l_expected,
                },
                Self::InvalidExitCode {
                    actual: r_actual,
                    expected: r_expected,
                },
            ) => l_actual == r_actual && l_expected == r_expected,
            (Self::InternalError(l0), Self::InternalError(r0)) => l0.to_string() == r0.to_string(),
            (_, _) => false,
        }
    }
}

impl Serialize for TestCaseError {
    fn serialize<S>(&self, serializer: S) -> anyhow::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self {
            Self::MalformedOutput(diff) => {
                let mut variant = serializer.serialize_map(Some(2))?;
                variant.serialize_entry("kind", "malformed_output")?;
                variant.serialize_entry("diff", &diff.lines)?;
                variant.end()
            }
            Self::InvalidExitCode { actual, expected } => {
                let mut variant = serializer.serialize_map(Some(3))?;
                variant.serialize_entry("kind", "invalid_exit_code")?;
                variant.serialize_entry("actual", actual)?;
                variant.serialize_entry("expected", expected)?;
                variant.end()
            }
            Self::InternalError(err) => {
                let mut variant = serializer.serialize_map(Some(2))?;
                variant.serialize_entry("kind", "internal_error")?;
                variant.serialize_entry("error", &format!("{}", err))?;
                variant.end()
            }
            Self::Skipped => {
                let mut variant = serializer.serialize_map(Some(1))?;
                variant.serialize_entry("kind", "skipped")?;
                variant.end()
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::TestCase;
    use super::TestCaseError;
    use crate::diff::Diff;
    use crate::diff::DiffLine;
    use crate::test_expectation;

    #[test]
    fn test_validate_succeeds_on_valid() {
        let testcase = TestCase {
            title: "an testcase".to_string(),
            shell_expression: "a command".to_string(),
            expectations: vec![test_expectation!("no-eol", "the stdout")],
            exit_code: Some(123),
            line_number: 234,
        };
        testcase
            .validate(&("the stdout", "the stderr", Some(123)).into())
            .expect("no error");
    }

    #[test]
    fn test_validate_fails_on_invalid_exit_code() {
        let testcase = TestCase {
            title: "an testcase".to_string(),
            shell_expression: "a command".to_string(),
            expectations: vec![test_expectation!("no-eol", "the stdout", false, false)],
            exit_code: Some(234),
            line_number: 123,
        };
        let asserted_output = ("the stdout", "the stderr", Some(123)).into();
        let result = testcase.validate(&asserted_output);
        match result {
            Ok(_) => panic!("assertion should have failed"),
            Err(err) => match err {
                TestCaseError::InvalidExitCode { actual, expected } => {
                    assert_eq!(
                        asserted_output.exit_code.as_code(),
                        actual,
                        "asserted output is delegated"
                    );
                    assert_eq!(234, expected, "expected exit code is delegated");
                }
                _ => panic!("unexpected error: {:?}", err),
            },
        }
    }

    #[test]
    fn test_validate_fails_on_malformed_output() {
        let testcase = TestCase {
            title: "an testcase".to_string(),
            shell_expression: "a command".to_string(),
            expectations: vec![test_expectation!(
                "no-eol",
                "something not matching",
                false,
                false
            )],
            exit_code: Some(123),
            line_number: 234,
        };
        let asserted_output = ("the stdout", "the stderr", Some(123)).into();
        let result = testcase.validate(&asserted_output);
        match result {
            Ok(_) => panic!("assertion should have failed"),
            Err(err) => {
                assert_eq!(
                    TestCaseError::MalformedOutput(Diff::new(vec![
                        DiffLine::UnmatchedExpectation {
                            index: 0,
                            expectation: testcase.expectations[0].clone()
                        },
                        DiffLine::UnexpectedLines {
                            lines: vec![(0, b"the stdout".to_vec())]
                        },
                    ])),
                    err,
                    "expected exit code is delegated"
                );
            }
        }
    }
}
