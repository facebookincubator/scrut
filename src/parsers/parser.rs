use std::fmt::Display;

use anyhow::Result;
use clap::ValueEnum;

use crate::testcase::TestCase;

/// A Parser extracts one or more [`crate::testcase::TestCase`]s from a provided text
pub trait Parser {
    /// Returns all testcases found in the provided text
    fn parse(&self, tests: &str) -> Result<Vec<TestCase>>;
}

#[derive(Debug, Clone, Copy, PartialEq, ValueEnum)]
pub enum ParserType {
    Markdown,
    Cram,
}

impl ParserType {
    pub fn file_extension(&self) -> &'static str {
        match self {
            Self::Cram => "t",
            Self::Markdown => "md",
        }
    }
}

impl Display for ParserType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Self::Cram => "cram",
                Self::Markdown => "markdown",
            }
        )
    }
}

impl TryFrom<String> for ParserType {
    type Error = String;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        match &value as &str {
            "markdown" | "md" => Ok(Self::Markdown),
            "cram" => Ok(Self::Cram),
            _ => Err(format!("Unsupported parser format `{value}`")),
        }
    }
}
