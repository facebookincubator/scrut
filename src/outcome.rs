/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::HashMap;

use serde::ser::SerializeMap;
use serde::Serialize;

use crate::escaping::Escaper;
use crate::output::Output;
use crate::parsers::parser::ParserType;
use crate::testcase::Result as TestCaseResult;
use crate::testcase::TestCase;

/// Aggregation of all that a renderer could possibly need to build a readable,
/// understandable output
pub struct Outcome {
    /// The path / URL of the test
    pub location: Option<String>,

    /// The output that this outcome describes
    pub output: Output,

    /// The testcase that this outcome is about
    pub testcase: TestCase,

    /// The original format the testcase was written in
    pub format: ParserType,

    // The output escaping mode this outcome needs to have
    pub escaping: Escaper,

    /// The result of validating the [`crate::testcase::TestCase`]
    pub result: TestCaseResult<()>,
}

impl Serialize for Outcome {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut count = 2;
        if self.location.is_some() {
            count += 1;
        }
        if self.result.is_err() {
            count += 1;
        }
        let mut outcome = serializer.serialize_map(Some(count))?;
        if let Some(ref location) = self.location {
            outcome.serialize_entry("location", location)?;
        }
        match &self.result {
            Err(err) => {
                outcome.serialize_entry("output", &self.output)?;
                outcome.serialize_entry("testcase", &self.testcase)?;
                outcome.serialize_entry("result", err)?;
            }
            Ok(_) => {
                outcome.serialize_entry("title", &self.testcase.title)?;
                let mut map = HashMap::new();
                map.insert("kind", "success");
                outcome.serialize_entry("result", &map)?;
            }
        }
        outcome.end()
    }
}

#[cfg(test)]
mod tests {
    use super::Outcome;
    use crate::escaping::Escaper;
    use crate::parsers::parser::ParserType;
    use crate::testcase::TestCaseError;

    #[test]
    fn test_serialize() {
        use crate::test_expectation;
        use crate::testcase::TestCase;

        let outcomes = vec![
            (
                "error",
                Outcome {
                    location: Some("path/file.md".to_string()),
                    output: ("stdout", "stderr", Some(123)).into(),
                    testcase: TestCase {
                        title: "the title".to_string(),
                        shell_expression: "the command".to_string(),
                        expectations: vec![test_expectation!("equal", "foo")],
                        exit_code: Some(234),
                        line_number: 234,
                        ..Default::default()
                    },
                    result: Err(TestCaseError::InvalidExitCode {
                        actual: 123,
                        expected: 234,
                    }),
                    escaping: Escaper::default(),
                    format: ParserType::Markdown,
                },
            ),
            (
                "success",
                Outcome {
                    location: Some("path/file.md".to_string()),
                    output: ("stdout", "stderr", Some(123)).into(),
                    testcase: TestCase {
                        title: "the title".to_string(),
                        shell_expression: "the command".to_string(),
                        expectations: vec![test_expectation!("equal", "foo")],
                        exit_code: Some(123),
                        line_number: 234,
                        ..Default::default()
                    },
                    result: Ok(()),
                    escaping: Escaper::default(),
                    format: ParserType::Markdown,
                },
            ),
        ];

        for (name, outcome) in outcomes {
            insta::assert_json_snapshot!(name, outcome);
        }
    }
}
