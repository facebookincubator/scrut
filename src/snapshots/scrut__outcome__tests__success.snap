---
source: src/outcome.rs
expression: outcome
---
{
  "location": "path/file.md",
  "title": "the title",
  "result": {
    "kind": "success"
  }
}
